### PR TITLE
[DRAFT]JsonExtensionData should allow non-JsonElement types during serialization

### PR DESCRIPTION
**Pull request by MarcoRossignoli**
working on https://github.com/dotnet/corefx/issues/38238

THIS IS A TMP DRAFT PR(INCOMPLETE) TO DISCUSS THE ISSUE, I'LL CLOSE.

Fix serialization it's not a great problem, the hard part is understand how to handle "deserialization" of numbers/object.
I tested Newtonsoft.Json and I would know if should be ok behave in the same way.
```cs
 public class TestModel
    {
        public string Name { get; set; } = "Test";

        [JsonExtensionDataAttribute]
        public Dictionary<string, object> Extensions { get; set; } = new Dictionary<string, object>()
        {
            // ["foo"] = 1.1
        };
    }

    class Program
    {
        static void Main(string[] args)
        {
            string s = @"{""Name"":""Test"",""foo"":1e-005}";
            var s1 = JsonConvert.DeserializeObject<TestModel>(s);
            Console.WriteLine(s1.Extensions["foo"].GetType());

            s = @"{""Name"":""Test"",""foo"":1.1}";
            s1 = JsonConvert.DeserializeObject<TestModel>(s);
            Console.WriteLine(s1.Extensions["foo"].GetType());

            s = @"{""Name"":""Test"",""foo"":1}";
            s1 = JsonConvert.DeserializeObject<TestModel>(s);
            Console.WriteLine(s1.Extensions["foo"].GetType());

            s = @"{""Name"":""Test"",""foo"":-1}";
            s1 = JsonConvert.DeserializeObject<TestModel>(s);
            Console.WriteLine(s1.Extensions["foo"].GetType());

            s = @"{""Name"":""Test"",""foo"":""text""}";
            s1 = JsonConvert.DeserializeObject<TestModel>(s);
            Console.WriteLine(s1.Extensions["foo"].GetType());

            s = @"{""Name"":""Test"",""foo"":{""prop"":""text""}}";
            s1 = JsonConvert.DeserializeObject<TestModel>(s);
            Console.WriteLine(s1.Extensions["foo"].GetType());

            s = @"{""Name"":""Test"",""foo"":true}";
            s1 = JsonConvert.DeserializeObject<TestModel>(s);
            Console.WriteLine(s1.Extensions["foo"].GetType());

            s = @"{""Name"":""Test"",""foo"":false}";
            s1 = JsonConvert.DeserializeObject<TestModel>(s);
            Console.WriteLine(s1.Extensions["foo"].GetType());
        }
    }
```
Result
```
System.Double
System.Double
System.Int64
System.Int64
System.String
Newtonsoft.Json.Linq.JObject <- we have JsonElement
System.Boolean
System.Boolean
```

@steveharter @ahsonkhan  before go on and prepare a complete PR I would like to know if I'm on right way or if there was another plan, thank's!

> Note that in the one use case that MVC uses this feature, it stores arbitrary content (strings or numbers) 

Or should we allow serialization and left deserialization to add only JsonElement on deserialization?

cc: @pranavkm

### CODE REVIEW

**Comment by pranavkm** [comment]
@MarcoRossignoli I wouldn't change the semantics of how deserialization occurs. The only thing that's necessary to have a good experience with the attribute is to remove this check: https://github.com/dotnet/corefx/blob/master/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleDictionary.cs#L141

**Comment by MarcoRossignoli** [comment]
@pranavkm I don't think that remove that check is enough...only entry value of type `JsonElement` has got `WriteAsProperty`, we should serialize in some way values, right?
I think that if entry value is object I think we need to serialize extension like "normal dictionary".
BTW if deserialization is not a problem(we deserialize all values as `JsonElement` so users will need to cast to it and call correct `Get[Type]`) update is less intrusive and the code is something similar
https://github.com/dotnet/corefx/blob/6c99cf5a0087bb83872c21a56b6bc33e9f4439ba/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleDictionary.cs#L43-L63

If you confirm that it's ok keep deserialization asis(all `JsonElement`) I can go on.

**Comment by pranavkm** [comment]
> If you confirm that it's ok keep deserialization asis(all JsonElement) I can go on.

That seems perfectly reasonable.

**Comment by ahsonkhan** [comment]
> Fix serialization it's not a great problem, the hard part is understand how to handle "deserialization" of numbers/object.

As @pranavkm mentioned, this should only be fixed for the serialization side. The deserialization should continue to be as JsonElement (and not other strongly typed types).

File: src/System.Text.Json/tests/Serialization/ExtensionDataTests.cs [hunk 1]
```diff
@@ -214,6 +184,49 @@ public static void ObjectTree()
...
+        }
+
+        [Fact]
+        public static void ExtensionPropertyDictionaryStringObject_Read()
+        {
+            ClassWithExtensionPropertyAsObject obj = JsonSerializer.Parse<ClassWithExtensionPropertyAsObject>(@"{""MyOverflow"":{""key"":1e-005}}");
+            Assert.IsType<double>(obj.MyOverflow["key"]);
+            Assert.Equal(1e-005, (double)obj.MyOverflow["key"]);
+
+            obj = JsonSerializer.Parse<ClassWithExtensionPropertyAsObject>(@"{""MyOverflow"":{""key"":1.1}}");
```
Comment (steveharter) [inline src/System.Text.Json/tests/Serialization/ExtensionDataTests.cs]: We do not want to "guess" here. We just want to create JsonElement on deserialize.

Comment (MarcoRossignoli) [reply src/System.Text.Json/tests/Serialization/ExtensionDataTests.cs]: Ok understood thanks

**Comment by MarcoRossignoli** [comment]
Thank's to all for fast feedback, close this for now because I get all answer I needed, tomorrow(CET) I'll PR final code.

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleDictionary.cs [hunk 2]
```diff
@@ -40,34 +40,27 @@ public static partial class JsonSerializer
                 // Handle DataExtension.
                 if (ReferenceEquals(jsonPropertyInfo, state.Current.JsonClassInfo.DataExtensionProperty))
```
Comment (steveharter) [inline src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleDictionary.cs]: I think we just need to remove lines 41-45 (and the WriteExtensionData method) + update tests.

Comment (MarcoRossignoli) [reply src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleDictionary.cs]: I opened another PR https://github.com/dotnet/corefx/pull/38306

**Pull request closed without merging**